components/toast_wrapper/index.tsx

```diff
@@ -2,37 +2,60 @@
 // See LICENSE.txt for license information.
 
 import {connect} from 'react-redux';
-import {bindActionCreators} from 'redux';
-import {withRouter} from 'react-router-dom';
+import {bindActionCreators, Dispatch} from 'redux';
+import {withRouter, RouteComponentProps} from 'react-router-dom';
+
+import {IDMappedObjects} from '@mattermost/types/utilities';
+
+import {Post} from 'packages/types/src/posts';
+
+import {GlobalState as ToastWrapperState} from '@mattermost/types/store';
+
+import {ViewsState} from 'types/store/views';
 
 import {createSelector} from 'reselect';
 
 import {Posts} from 'mattermost-redux/constants';
+
 import {getAllPosts, getPostIdsInChannel} from 'mattermost-redux/selectors/entities/posts';
+
 import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
+
 import {makePreparePostIdsForPostList} from 'mattermost-redux/utils/post_list';
+
+import {Channel} from '@mattermost/types/channels';
+
 import {getCurrentChannel, countCurrentChannelUnreadMessages, isManuallyUnread} from 'mattermost-redux/selectors/entities/channels';
+
 import {getUnreadScrollPositionPreference, isCollapsedThreadsEnabled} from 'mattermost-redux/selectors/entities/preferences';
 
 import {updateToastStatus} from 'actions/views/channel';
 
-import ToastWrapper from './toast_wrapper.jsx';
+import ToastWrapper from './toast_wrapper';
+
+type GlobalState = ToastWrapperState & {
+    views: ViewsState;
+}
+
+type OwnProps = RouteComponentProps & {
+    channelId: string;
+    atLatestPost?: boolean;
+}
+
 export function makeGetRootPosts() {
     return createSelector(
         'makeGetRootPosts',
         getAllPosts,
         getCurrentUserId,
         getCurrentChannel,
-        (allPosts, currentUserId, channel) => {
+        (allPosts: IDMappedObjects<Post>, currentUserId: string, channel: Channel) => {
             // Count the number of new posts that haven't been deleted and are root posts
-            return Object.values(allPosts).filter((post) => {
-                return (
-                    post.root_id === '' &&
-                    post.channel_id === channel.id &&
-                    post.state !== Posts.POST_DELETED
-                );
+            return Object.values(allPosts).filter((post: Post) => {
+                return (post.root_id === '' &&
+                post.channel_id === channel.id &&
+                post.state !== Posts.POST_DELETED);
             }).reduce((map, obj) => {
```

Comment:
type of `map` should be `Record<string, boolean>`. Type of `rootPosts` should be adjusted accordingly in `toast_wrapper.tsx`

```suggestion
            }).reduce((map: Record<string, boolean>, obj) => {
```